staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation_test.go

```diff
@@ -458,3 +461,112 @@ func TestValidateCustomResourceDefinitionUpdate(t *testing.T) {
 		}
 	}
 }
+
+func TestValidateCustomResourceDefinitionOpenAPISchema(t *testing.T) {
+	t.Parallel()
+	tests := []struct {
+		name     string
+		resource *apiextensions.CustomResourceDefinition
+		errors   []validationMatch
+	}{
+		{
+			name: "valid 1",
+			resource: &apiextensions.CustomResourceDefinition{
+				ObjectMeta: metav1.ObjectMeta{Name: "plural.group.com"},
+				Spec: apiextensions.CustomResourceDefinitionSpec{
+					Group:   "group.com",
+					Version: "version",
+					Names: apiextensions.CustomResourceDefinitionNames{
+						Plural:   "plural",
+						Singular: "singular",
+						Kind:     "kind",
+						ListKind: "listkind",
+					},
+					Scope: apiextensions.ClusterScoped,
+					Validation: &apiextensions.CustomResourceValidation{
+						OpenAPIV3Schema: &apiextensions.JSONSchemaProps{
+							AdditionalProperties: &apiextensions.JSONSchemaPropsOrBool{
+								Allows: true,
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "valid 2",
+			resource: &apiextensions.CustomResourceDefinition{
+				ObjectMeta: metav1.ObjectMeta{Name: "plural.group.com"},
+				Spec: apiextensions.CustomResourceDefinitionSpec{
+					Group:   "group.com",
+					Version: "version",
+					Names: apiextensions.CustomResourceDefinitionNames{
+						Plural:   "plural",
+						Singular: "singular",
+						Kind:     "kind",
+						ListKind: "listkind",
+					},
+					Scope: apiextensions.ClusterScoped,
+					Validation: &apiextensions.CustomResourceValidation{
+						OpenAPIV3Schema: &apiextensions.JSONSchemaProps{
+							AdditionalProperties: &apiextensions.JSONSchemaPropsOrBool{
+								Schema: &apiextensions.JSONSchemaProps{},
```

Comment:
This is what I'm trying to do but it does not work.

Reply:
While this PR makes perfect sense to me, I am a little confused if this really conforms to the Kubernetes API conventions.

When we specify a schema in `additionalProperties`, it means "the new fields are only allowed if they conform according to this schema". This makes sense in a case where you want to validate the new fields. However, this also means that you are not ignoring new fields (which goes against Kubernetes API conventions).

@sttts @ash2k wdyt?

Reply:
I'm trying to describe restrictions for labels/annotations in the schema.

https://github.com/atlassian/smith/blob/5a992f721782c738f34201d3d74b60e0e11e221b/pkg/resources/crd_helpers.go#L77-L96

Seems like a perfectly valid use case :) If the schema author knows in advance that ALL elements in the object must conform to the spec then... this is not friendly to adding new fields later that don't. It can be expressed in a backward compatible way in the future but it cannot be made forward compatible.